ui/components/ui/currency-input/currency-input.js

```diff
@@ -0,0 +1,166 @@
+import PropTypes from 'prop-types';
+import UnitInput from '../unit-input';
+import CurrencyDisplay from '../currency-display';
+import {
+  getValueFromWeiHex,
+  getWeiHexFromDecimalValue,
+} from '../../../helpers/utils/conversions.util';
+import { ETH } from '../../../helpers/constants/common';
+import { I18nContext } from '../../../contexts/i18n';
+
+/**
+ * Component that allows user to enter currency values as a number, and props receive a converted
+ * hex value in WEI. props.value, used as a default or forced value, should be a hex value, which
+ * gets converted into a decimal value depending on the currency (ETH or Fiat).
+ */
+export default function CurrencyInput({
+  hexValue,
+  preferredCurrency,
+  secondaryCurrency,
+  hideSecondary,
+  featureSecondary,
+  conversionRate,
+  primarySuffix,
+  secondarySuffix,
+  onChange,
+  onPreferenceToggle,
+}) {
+  const t = useContext(I18nContext);
+
+  const [isSwapped, setSwapped] = useState(false);
+  const [newHexValue, setNewHexValue] = useState(hexValue);
+
+  const shouldUseFiat = () => {
+    if (hideSecondary) {
+      return false;
+    }
+
+    return Boolean(featureSecondary);
+  };
+
+  const getDecimalValue = () => {
+    const decimalValueString = shouldUseFiat()
+      ? getValueFromWeiHex({
+          value: hexValue,
+          toCurrency: secondaryCurrency,
+          conversionRate,
+          numberOfDecimals: 2,
+        })
+      : getValueFromWeiHex({
+          value: hexValue,
+          toCurrency: ETH,
+          numberOfDecimals: 8,
+        });
+
+    return Number(decimalValueString) || 0;
+  };
+
+  const initialDecimalValue = hexValue ? getDecimalValue() : 0;
+  const [decimalValue, setDecimalValue] = useState(initialDecimalValue);
+
+  useEffect(() => {
+    setNewHexValue(hexValue);
+    const newDecimalValue = getDecimalValue();
+    setDecimalValue(newDecimalValue);
+  }, [hexValue]);
+
+  const swap = async () => {
+    await onPreferenceToggle(!featureSecondary);
+    setSwapped(!isSwapped);
+  };
+
+  const handleChange = (newDecimalValue) => {
+    const hexValueNew = shouldUseFiat()
+      ? getWeiHexFromDecimalValue({
+          value: newDecimalValue,
+          fromCurrency: secondaryCurrency,
+          conversionRate,
+          invertConversionRate: true,
+        })
+      : getWeiHexFromDecimalValue({
+          value: newDecimalValue,
+          fromCurrency: ETH,
+          fromDenomination: ETH,
+          conversionRate,
+        });
+
+    setNewHexValue(hexValueNew);
+    setDecimalValue(newDecimalValue);
+    onChange(hexValueNew);
+    setSwapped(!isSwapped);
+  };
+
+  useEffect(() => {
+    if (isSwapped) {
+      handleChange(decimalValue);
+    }
+  }, [isSwapped]);
+
+  const renderConversionComponent = () => {
+    let currency, numberOfDecimals;
+
+    if (hideSecondary) {
+      return (
+        <div className="currency-input__conversion-component">
+          {t('noConversionRateAvailable')}
```

Comment:
related comment on the issue ticket here: https://github.com/MetaMask/metamask-extension/issues/12601#issuecomment-985680398

Reply:
@adonesky1 I think that comment is correct, but it should probably be addressed in a separate PR. This PR is just maintaining the current behaviour (currently implemented in https://github.com/MetaMask/metamask-extension/pull/12813/files#diff-6afdf120de27d111574e2c0e237c729a4218101d99419d68773c1c633731323dL126-L132) as it refactors the component